### ISSUE TITLE
Svelte6 readiness: Why are quotes required for Pug attributes?

### ISSUE THREAD

**Issue by njreid** [sveltejs/svelte-preprocess#660]
**Describe the bug**
Currently, Pug syntax embedded in Svelte template tags requires quoted values. 

`
body
  p(class:name!="{highlighted}") Example
`

This isn't a Pug requirement. The following Pug syntax would be fine:

`
body
  p(class:name={highlighted} Example
`

In Svelte 6, I understand that quoted attribute will be stringified. So, the way Svelte interprets the HTML equivalent of the previous two examples will be different.

What is the reason that the quotes are currently required? Can we remove that requirement in svelte-preprocess?

**Comment by Conduitry** [comment]
What happens if the expression has spaces in it, or any other special characters? How does this get parsed then?

**Comment by njreid** [comment]
Any spaces between the braces should be ok, right? 

```pug
p(class:name={something.length > 0}) Example Test
```

**Comment by CosmoMyzrailGorynych** [comment]
> Any spaces between the braces should be ok, right?

Incorrect. The `{}` combination itself is read by Pug as an object — because yes, `{}` is a valid JavaScript expression, and Pug evaluates these expressions when compiling templates. (See [Attributes](https://pugjs.org/language/attributes.html), the second paragraph.) With that, the compiler fails on `.` in `something.length > 0` because it is not allowed in keys in the usual object notation.

**Comment by CosmoMyzrailGorynych** [comment]
@njreid your example in the issue's description is also incorrect.

<sup>And also has a syntax error.</sup>

So yes, it is a requirement for Pug to quote special characters, which include those in Svelte directives and templated attributes, because Svelte consumes HTML produced by Pug, not Pug directly.

All in all I think that the quotes behavior in Svelte isn't well-thought (Pug is not the only HTML preprocessor), especially when it avoids "magical" behavior now. (The presence of quotes is too invisible to be an obvious syntax feature for the desired language curve.) And the stupid part is that there is seemingly no alternative or option to change this behavior.

---
Posting a relevant error message so this issue pops up in searches:
```
▲ [WARNING] Quoted attributes on components and custom elements will be stringified in a future version of Svelte. If this isn't what you want, remove the quotes
https://svelte.dev/e/attribute_quoted [plugin esbuild-svelte]
```

**Comment by njreid** [comment]
> [@njreid](https://github.com/njreid) your example in the issue's description is also incorrect.
> 
> And also has a syntax error.
> 
> So yes, it is a requirement for Pug to quote special characters, which include those in Svelte directives and templated attributes, because Svelte consumes HTML produced by Pug, not Pug directly.
> 
> All in all I think that the quotes behavior in Svelte isn't well-thought (Pug is not the only HTML preprocessor), especially when it avoids "magical" behavior now. (The presence of quotes is too invisible to be an obvious syntax feature for the desired language curve.) And the stupid part is that there is seemingly no alternative or option to change this behavior.
> 
> Posting a relevant error message so this issue pops up in searches:
> 
> ```
> ▲ [WARNING] Quoted attributes on components and custom elements will be stringified in a future version of Svelte. If this isn't what you want, remove the quotes
> https://svelte.dev/e/attribute_quoted [plugin esbuild-svelte]
> ```

Thanks - fixed the missing paren in the original example.

**Comment by i18nsite** [comment]
## A Working Solution: Pre-process + Post-process in Custom Pug Handler

I ran into the same issue and found a practical workaround that eliminates all `attribute_quoted` warnings while keeping Pug as the template language. The approach has two parts:

### The Problem

Pug always outputs quoted attribute values in HTML. So `attr!="{expr}"` in Pug source compiles to `attr="{expr}"` in HTML, which Svelte 5 warns about (and Svelte 6 will break).

### The Solution

Instead of modifying `svelte-preprocess` itself, I use a custom Pug preprocessor with two steps:

**Step 1: Pre-process Pug source** — Allow writing `attr={expr}` (unquoted) in Pug templates. Before Pug compilation, convert these to `attr!="{expr}"` so Pug can parse them:

```js
// In the Pug attribute parser (split function), track {} depth so
// attr={expr} is treated as a single token rather than being split at spaces:

// state 2: inside curly braces
if (i === '{') { state = 2; depth = 1; t.push(i); continue; }
// ...
} else if (state === 2) {
  if (i === '{') depth++;
  else if (i === '}') depth--;
  t.push(i);
  if (depth === 0) state = 0;
}

// Then in attribute processing, detect attr={expr} pattern and wrap with quotes:
if (eq_pos > 0 && attr.charAt(eq_pos + 1) === '{') {
  const pre = attr.slice(0, eq_pos), val = attr.slice(eq_pos + 1);
  if (pre.endsWith('!')) {
    // attr!={expr} → attr!="{expr}" (already has !, don't double it)
    set(pre + '="' + val + '"');
  } else {
    // attr={expr} → attr!="{expr}" (add != for unescaped output)
    wrap(pre + '=', val);
  }
}
```

**Step 2: Post-process compiled HTML** — After Pug compiles to HTML, strip quotes from pure expression attributes:

```js
// In the Svelte markup preprocessor, after pug2htm compilation:
const render = (content, filename, dir) =>
  pug2htm(pugHack(content), filename, { ... })
    .code.replace(/="\{([^"]*)\}"/g, '={$1}');
```

This regex converts `attr="{expr}"` → `attr={expr}` in the compiled HTML, which is exactly what Svelte expects.

### Result

- ✅ All `attribute_quoted` warnings eliminated
- ✅ Supports both quoted `attr="{expr}"` and unquoted `attr={expr}` syntax in Pug
- ✅ Handles `attr!={expr}` (Pug's unescaped output) correctly
- ✅ Zero breaking changes — existing `attr="{expr}"` style continues to work

### Usage in Pug templates

```pug
//- New unquoted style (recommended)
i-my-component(msg={i[2]} self!={i[0]===1})

//- Old quoted style (still works)
i-my-component(msg="{i[2]}" self!="{i[0]===1}")
```

---

## 中文说明

### 问题

Pug 编译器输出的 HTML 属性值总是带引号，比如 `attr="{expr}"`。在 Svelte 5 中这会触发 `attribute_quoted` 警告，而在 Svelte 6 中带引号的属性会被直接字符串化，导致功能异常。

### 解决方案

我使用自定义的 Pug 预处理器，通过**预处理 + 后处理**两步解决：

1. **预处理（编译前）**：允许在 Pug 模板中使用 `attr={expr}` 无引号写法。在 Pug 编译前，自动将 `attr={expr}` 转为 `attr!="{expr}"`（加引号让 Pug 能正确解析）。关键是让属性解析器（split 函数）识别花括号 `{}` 作为完整的值范围，不在花括号内因空格而拆分。同时区分 `=` 和 `!=` 两种前缀，避免产生 `!!=` 双感叹号的错误。

2. **后处理（编译后）**：Pug 编译出的 HTML 中，用正则 `="\{([^"]*)\}"/g` 匹配纯表达式属性（整个值是一个 `{...}` 的），去掉引号变为 `={expr}`。这样 Svelte 收到的就是不带引号的属性，不再产生警告。

### 效果

- 所有 `attribute_quoted` 警告消除
- 支持新旧两种写法：`attr={expr}`（推荐）和 `attr="{expr}"`（兼容）
- 对带文本拼接的属性（如 `attr="text {expr}"`）不受影响
